---
title: "Next and 3rd generation sequencing adventures"
author: "Armand Bester"
format: 
    revealjs:
        theme: default
---


# Infant microbiome

::: {#fig-alpha layout-ncol=2}
![](media/infant_alpha.png)

![](media/milton.png)

Alpha diversities over time
:::

$$
H = - \sum_{i=1}^k p_i log_2(p_i)
$${#eq-Shannon}

## ...Example results

* [Python notebook](html/Kraken_and_bracken.html)


* Krona reports
    
    [Krona 5b](html/krona/VRM-5B.krona.html)

    [Krona 5d](html/krona/VRM-5D.krona.html)

* [Pavian analysis](html/Uploaded_sample_set-report.html)



# A pandemic happens

::: {#fig-pandemic layout-ncol=1}

![](media/sars2.png){width=30%}

 SARS-COV2 rendering Dall-E 2
:::

## Solutions

[Roll your own](media/ryo_depth.png)

::: {#fig-roll_your_own layout-ncol=1}

![](media/roll_y_own.png){width=80%}

:::


## ... Don't reinvent the wheel

[Smarter people](html/ONT%20sequencing.html)


::: {#fig-artic layout-ncol=1}

![](media/ARTIC-ncov.png){width=60%}

 [https://artic.network/ncov-2019/ncov2019-bioinformatics-sop.html](https://artic.network/ncov-2019/ncov2019-bioinformatics-sop.html)

:::

## ...

[Minimum effort](html/wf-artic-040822_ONT%20(copy).html)

::: {#fig-epi2me layout-ncol=1}

![](media/midnight.png){width=60%}

 [https://labs.epi2me.io/sarscov2-midnight-analysis/](https://labs.epi2me.io/sarscov2-midnight-analysis/)

:::

## ... mapped reads

::: {#fig-igv layout-ncol=1}

![](media/sars_cov2_igv_snapshot.png)

An IGV snapshot
:::

## ... small part of a global effort


::: {#fig-global layout-ncol=1}

![](media/sars-cov2-global.png){width=100%}

 [https://nextstrain.org/ncov/gisaid/global/6m](https://nextstrain.org/ncov/gisaid/global/6m)

:::


## ...


::: {#fig-primal_scheme layout-ncol=1}

![](media/primal_scheme.png){width=100%}

 [https://primalscheme.com](https://primalscheme.com)

:::


## ... and having fun


::: {#fig-primal_scream layout-ncol=1}

![](media/primal_scream.png){width=60%}

 [https://www.rs500albums.com/450-401/437](https://www.rs500albums.com/450-401/437)

:::


## ... even mobile

::: {#fig-primal_mobile layout-ncol=1}

![](media/mobile.png){width=55%}

 [ Nature (2016) https://doi.org/10.1038/nature16996](https://www.nature.com/articles/nature16996)

:::



# Mozzie hitchhikers

::: {#fig-mosquito}
![](media/DALLE_mosquito.png)

Mosquito rendering Dall-E 2
:::

## 'Bin' reads and BLAST

::: {#fig-amplicon_sorter}

![](media/amplicon_sorter.png)

[https://github.com/avierstr/amplicon_sorter](https://github.com/avierstr/amplicon_sorter)
:::


## Sanity plots

```{python}
#| label: fig-id
#| fig-cap: "% identity"

import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import plotly.express as px


names = "qaccver saccver pident bitscore evalue length mismatch gapopen qstart qend sstart send ssciname stitle staxid scomname sblastname sskingdom staxids".split(" ")

blstDbAll = pd.read_csv("data/ampSorter_blst_nt.csv", delimiter="\t", names=names)
maxBitScr = blstDbAll.sort_values('bitscore', ascending=False).groupby('qaccver', as_index=False).first()
maxBitScr['BC'] = maxBitScr['qaccver'].apply(lambda s: s.split("_")[1])
maxBitScr['N'] = maxBitScr['qaccver'].apply(lambda s: int(s.split("(")[1].replace(")", "")))
maxBitScr['Total'] = maxBitScr.groupby(['BC'])['N'].transform('sum')
maxBitScr['Ng'] = maxBitScr.groupby(['BC', 'ssciname'])['N'].transform('sum')
maxBitScr.drop_duplicates(subset=['BC', 'ssciname'], inplace=True, keep='first')
maxBitScr['Frac'] = maxBitScr['Ng'] / maxBitScr['Total'] * 100


sns.boxplot(data=maxBitScr, x='pident')
plt.title("A boxplot of the overall % ID match");
```


---

```{python}
#| label: fig-eval
#| fig-cap: "e-value"

sns.boxplot(data=maxBitScr, x='evalue')
plt.title("A boxplot of the overall e-value");
```

---

```{python}
#| label: fig-count
#| fig-cap: "Count plot"

sns.countplot(y=maxBitScr['sskingdom']);
```

---


```{python}
#| label: fig-readCov
#| fig-cap: "Read coverage"

plt.bar(x=maxBitScr['BC'], height=maxBitScr['Total'])
plt.xticks(rotation=90)
plt.hlines(y=200, xmin=-1, xmax=24, color='red')
plt.text(x=23.5, y=300, s='200', color='red')
plt.title("Reads per barcode");
```


## What did we find?

```{python}
#| label: fig-interactive-id
#| fig-cap: "Amplicon bins"

px.scatter(data_frame=maxBitScr, x='BC', y='ssciname', 
           color='sskingdom', size='Frac', height=400, 
          hover_data = ['N'])
```




# ITS with a focus on fungal diversity

::: {#fig-fungi}
![](media/fungi.png)

Fungi rendering Dall-E 2
:::


## ... Ascomycetes example

::: {#fig-asco}
![](media/asco.png)

Fungi rendering Dall-E 2
:::

[Example notebook](html/Blast_raw_reads_against_nt.html)


# Conclusion

* Bioinformatics is a broad field

* Many opportunities for adventures

* Requires some knowledge 
    * Linux / BASH
    * Python / R
    * A little math and a little stats

* In stead of clicking next go have a beer
    * Notebooks/script and loops makes for better reproducibility 


# Thank you

Questions?
